src/renderer/vt/Xterm256Engine.cpp

```diff
@@ -42,11 +47,74 @@ Xterm256Engine::Xterm256Engine(_In_ wil::unique_hfile hPipe,
     // We have to do this here, instead of in PaintBufferGridLines, because
     //      we'll have already painted the text by the time PaintBufferGridLines
     //      is called.
+    // TODO:GH#2915 Treat underline separately from LVB_UNDERSCORE
     RETURN_IF_FAILED(_UpdateUnderline(legacyColorAttribute));
 
+    // Only do extended attributes in xterm-256color, as to not break telnet.exe.
+    RETURN_IF_FAILED(_UpdateExtendedAttrs(extendedAttrs));
+
     return VtEngine::_RgbUpdateDrawingBrushes(colorForeground,
                                               colorBackground,
-                                              isBold,
+                                              WI_IsFlagSet(extendedAttrs, ExtendedAttributes::Bold),
                                               _ColorTable,
                                               _cColorTable);
 }
+
+// Routine Description:
+// - Write a VT sequence to either start or stop underlining text.
+// Arguments:
+// - legacyColorAttribute: A console attributes bit field containing information
+//      about the underlining state of the text.
+// Return Value:
+// - S_OK if we succeeded, else an appropriate HRESULT for failing to allocate or write.
+[[nodiscard]] HRESULT Xterm256Engine::_UpdateExtendedAttrs(const ExtendedAttributes extendedAttrs) noexcept
+{
+    // Helper lambda to check if a state (attr) has changed since it's last
+    // value (lastState), and appropriately start/end that state with the given
+    // begin/end functions.
+    auto updateFlagAndState = [extendedAttrs, this](const ExtendedAttributes attr,
+                                                    bool& lastState,
+                                                    std::function<HRESULT(Xterm256Engine*)> beginFn,
+                                                    std::function<HRESULT(Xterm256Engine*)> endFn) -> HRESULT {
+        const bool flagSet = WI_AreAllFlagsSet(extendedAttrs, attr);
+        if (flagSet != lastState)
+        {
+            if (flagSet)
+            {
+                RETURN_IF_FAILED(beginFn(this));
+            }
+            else
+            {
+                RETURN_IF_FAILED(endFn(this));
+            }
+            lastState = flagSet;
+        }
+        return S_OK;
+    };
+
+    auto hr = updateFlagAndState(ExtendedAttributes::Italics,
+                                 _usingItalics,
+                                 &Xterm256Engine::_BeginItalics,
+                                 &Xterm256Engine::_EndItalics);
+    RETURN_IF_FAILED(hr);
+
+    hr = updateFlagAndState(ExtendedAttributes::Blinking,

```

Comment:
Nnnn... what happens in the scenario where:

- We're attempting to set Italics, Blink, and Invisible.
- Italics succeeds
- Blink fails
- Invisible would have succeeded, but we early returned...

Are we OK with that happening or should we have a provision to try the others and aggregate the result code later?

I don't think there's really a way for us to make this atomic where they all work or none of them work... or is there?

EDIT: I'm worried that the potentially torn state will result in a weird debugging situation in the future...

Reply:
You're right that's it's possible that we'd end in a bad state if one of these fails. Looking at the implementation of these, this is the only spot that can actually fail:
```c++
    try
    {
        _buffer.append(str);
        return S_OK;
    }
    CATCH_RETURN();
```
(where `str` is the VT sequence we're writing). IMO, if that fails, then we're probably going to end up in a much worse torn state than just this one sequence being lost.

The idea to do them all atomically would probably match well with #3016.

Reply:
Let's stash the idea inside the notes for #3016 then and just let this roll as is.